src/db/spanner/batch.rs

```diff
@@ -269,22 +253,74 @@ pub async fn do_append_async(
-    let rows: Vec<_> = bsos
-        .into_iter()
-        .map(|bso| {
-            let sortindex = bso
-                .sortindex
-                .map(|sortindex| as_value(sortindex.to_string()))
-                .unwrap_or_else(null_value);
-            let payload = bso.payload.map(as_value).unwrap_or_else(null_value);
-            if payload != null_value() {
-                running_size += payload.get_string_value().len();
-            }
-            let ttl = bso
-                .ttl
-                .map(|ttl| as_value(ttl.to_string()))
-                .unwrap_or_else(null_value);
 
+    // problem: Append may try to insert a duplicate record into the batch_bsos table.
+    // this is because spanner doesn't do upserts. at all. Batch_bso is a temp table and
+    // items are eventually rolled into bsos.
+
+    fn exist_idx(collection_id: &str, batch_id: &str, bso_id: &str) -> String {
+        format!(
+            "{collection_id}::{batch_id}::{bso_id}",
+            collection_id = collection_id,
+            batch_id = batch_id,
+            bso_id = bso_id,
+        )
+    }
+
+    struct UpdateRecord {
+        sortindex: Value,
+        ttl: Value,
+        bso_id: String,
+        payload: String,
+    };
+
+    //prefetch the existing batch_bsos for this user.
+    let mut existing = HashSet::new();
+    let mut existing_stream = db
+    .sql("SELECT collection_id, batch_id, batch_bso_id from batch_bsos where fxa_uid=@fxa_uid and fxa_kid=@fxa_kid;")?
+    .params(params!{
+        "fxa_uid" => user_id.fxa_uid.clone(),
+        "fxa_kid" => user_id.fxa_kid.clone(),
+    }).execute_async(&db.conn)?;
+    while let Some(row) = existing_stream.next_async().await {
+        let row = row?;
+        existing.insert(exist_idx(
+            row[0].get_string_value(),
+            row[1].get_string_value(),
+            row[2].get_string_value(),
+        ));
+    }
+
+    // Approach 1:
+    // iterate and check to see if the record is in batch_bso table already
+    let mut insert: Vec<Value> = Vec::new();
+    let mut update: Vec<UpdateRecord> = Vec::new();
+    for bso in bsos {
+        let sortindex = bso
+            .sortindex
+            .map(|sortindex| as_value(sortindex.to_string()))
+            .unwrap_or_else(null_value);
+        let payload = bso.payload.map(as_value).unwrap_or_else(null_value);
+        if payload != null_value() {
+            running_size += payload.get_string_value().len();
+        }
+        let ttl = bso
+            .ttl
+            .map(|ttl| as_value(ttl.to_string()))
+            .unwrap_or_else(null_value);
+
+        let exist_idx = exist_idx(&collection_id.to_string(), &batch.id, &bso.id);
+
+        if existing.contains(&exist_idx) {
+            // need to update this record
+            update.push(UpdateRecord {
+                sortindex,
+                ttl,
+                bso_id: bso.id,
+                payload: payload.get_string_value().to_owned(),
```

Comment:
You don't want to copy here, especially payload
```suggestion
                payload: payload.take_string_value(),
```